tmk_core/common/action_util.c

```diff
@@ -279,3 +280,120 @@ uint8_t has_anymod(void)
+ *
+ * This function is meant to be used in process_record_user to be able to
+ * trigger specific action on key code with modifier, for example when shift-esc
+ * is pressed.
+ *
+ * \param modifier_mask modifier mask to check against current modifiers held.
+ * \param layer layer to check against current layer.
+ * \param keycode the processed keycode, will be forwarded to handler.
+ * \param record the processed record, will be forwarded to handler.
+ * \param data anonymous data to be passed to the handler.
+ * \param handler to be called when modifiers and layer are matched.
+ * \return true if modifiers or layer did not match, return value of handler otherwise.
+ */
+bool run_on_modifiers(uint8_t modifier_mask, int16_t layer, uint16_t keycode,  keyrecord_t *record,
+                      void *data,  bool (*handler)(uint16_t keycode,  keyrecord_t *record, void *data)) {
+  static bool in_alternate_modifier;
+
+  /* nothing to do if we are not in the right layer */
+  if (layer >= 0 && !layer_state_is(layer))
+    return true;
+
+  /* when it's a key press and modifier state is pressed */
+  if (record->event.pressed && (get_mods() & modifier_mask) == modifier_mask) {
+    in_alternate_modifier = true;
+
+    /* will send modifier up so that the os won't shift the keycode we will send */
+    del_mods(modifier_mask);
+
+    /* send mods modifications */
+    send_keyboard_report();
+
+    /* execute handler */
+    bool rc = handler(keycode, record, data);
+
+    add_mods(modifier_mask);
+
+    return rc;
+  }
+
+  /* when releasing the key and we activated alternate modifier */
+  if (!record->event.pressed && in_alternate_modifier) {
+    in_alternate_modifier = false;
+
+    /* execute handler */
+    bool rc = handler(keycode, record, data);
+
+    /* make sure all mods we sat up earlier are released */
+    clear_weak_mods();
+
+    /* send mods modification */
+    send_keyboard_report();
+
+    return rc;
+  }
+
+  return true;
+}
+
+/** \brief add a weak mod if keycode has the mod enabled */
+#define maybe_add_weak_mods(keycode, mod)                               \
+  if (keycode < QK_MODS_MAX &&                                          \
+      (keycode & 0xff00) == QK_ ## mod)                                 \
+    add_weak_mods(MOD_BIT(KC_ ## mod))
+
+/** \brief copy keycode mods as weak mod */
+static void copy_mods_as_weak(uint16_t keycode) {
+    /* send mods if keycode needs it */
+    maybe_add_weak_mods(keycode, LCTL);
+    maybe_add_weak_mods(keycode, LSFT);
+    maybe_add_weak_mods(keycode, LALT);
+    maybe_add_weak_mods(keycode, LGUI);
+    maybe_add_weak_mods(keycode, RCTL);
+    maybe_add_weak_mods(keycode, RSFT);
+    maybe_add_weak_mods(keycode, RALT);
+    maybe_add_weak_mods(keycode, RGUI);
+}
+
+/** \brief run_on_modifiers handler to send custom keycode
+ */
+static bool custom_keycode_on_modifiers_handler(uint16_t keycode, keyrecord_t *record, void *data) {
+  uint16_t custom_keycode =  (uint16_t) data;
```

Comment:
Looks like it doesn't like this:
```travis
tmk_core/common/action_util.c: In function ‘custom_keycode_on_modifiers_handler’:
tmk_core/common/action_util.c:371:30: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
   uint16_t custom_keycode =  (uint16_t) data;

tmk_core/common/action_util.c: In function ‘custom_keycode_on_modifiers’:
tmk_core/common/action_util.c:398:60: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
   return run_on_modifiers(modifier_mask, layer, 0, record, (void *) custom_keycode, custom_keycode_on_modifiers_handler);

```

Reply:
More than likely, these both should be uint16_t's and not voids.

Reply:
Hello, the goal here is to allow user of the run_on_modifiers function to pass arbitrary data so that's why I choose a void pointer here. I'll fix the user function to use a real pointer instead, unless you see a better way.

Reply:
Done, in fact the data to be passed to the handler can always be allocated on the stack so simply passing a pointer to it should do.